src/frequenz/sdk/microgrid/component_graph.py

```diff
@@ -398,18 +400,16 @@ def components(
             Set of the components currently connected to the microgrid, filtered by
                 the provided `component_ids` and `component_categories` values.
         """
-        if component_ids is None:
-            # If any node has not node[1], then it will not pass validations step.
-            selection: Iterable[Component] = map(
-                lambda node: Component(**(node[1])), self._graph.nodes(data=True)
-            )
-        else:
-            valid_ids = filter(self._graph.has_node, component_ids)
-            selection = map(lambda idx: Component(**self._graph.nodes[idx]), valid_ids)
+        selection: Iterable[Component]
+        selection_ids = (
+            self._graph.nodes
+            if component_ids is None
+            else component_ids & self._graph.nodes
+        )
+        selection = (self._graph.nodes[i][_DATA_KEY] for i in selection_ids)
```

Comment:
Optionally you might want to set the type and assign the value at the same time. And the definition of `selection` above can be deleted
```suggestion
        selection: Iterable[Component] = (self._graph.nodes[i][_DATA_KEY] for i in selection_ids)
```

Reply:
Yeah, declaring the type is not even necessary anymore, I think it was required by a previous iteration of this refactoring.

Reply:
Oh, no, it is still needed.